### ISSUE TITLE
ai: Develop and Train the GNN Model

### ISSUE THREAD

**Issue by sanamoin5** [sanamoin5/gadgets#5]
**Objective**: Create and train a Graph Neural Network to predict healthcare accessibility scores.

**Tasks**:

- **Set Up PyTorch Geometric Environment**:
  - Ensure all dependencies are installed:
    ```bash
    pip install torch torch-geometric
    ```
  - Verify installation by running a sample GNN model.
  
- **Implement Baseline Graph Convolutional Network (GCN)**:
  - **Define Model Architecture**:
    ```python
    import torch
    import torch.nn.functional as F
    from torch_geometric.nn import GCNConv

    class GCN(torch.nn.Module):
        def __init__(self, num_node_features, hidden_channels):
            super(GCN, self).__init__()
            self.conv1 = GCNConv(num_node_features, hidden_channels)
            self.conv2 = GCNConv(hidden_channels, 1)  # Predicting accessibility score

        def forward(self, data):
            x, edge_index = data.x, data.edge_index
            x = self.conv1(x, edge_index)
            x = F.relu(x)
            x = F.dropout(x, training=self.training)
            x = self.conv2(x, edge_index)
            return x
    ```
  
- **Prepare Training and Testing Data**:
  - **Data Splitting**:
    - Split regions into training and testing sets (e.g., 80% train, 20% test).
    - Example:
      ```python
      from sklearn.model_selection import train_test_split

      train_indices, test_indices = train_test_split(range(data.num_nodes), test_size=0.2)
      train_mask = torch.zeros(data.num_nodes, dtype=torch.bool)
      test_mask = torch.zeros(data.num_nodes, dtype=torch.bool)
      train_mask[train_indices] = True
      test_mask[test_indices] = True
      data.train_mask = train_mask
      data.test_mask = test_mask
      ```
  
  - **Normalize Features**:
    - Scale numerical features for better model performance.
    - Example using `scikit-learn`:
      ```python
      from sklearn.preprocessing import StandardScaler

      scaler = StandardScaler()
      data.x = torch.tensor(scaler.fit_transform(data.x.numpy()), dtype=torch.float)
      ```
  
- **Train the GNN Model**:
  - **Define Loss Function and Optimizer**:
    ```python
    criterion = torch.nn.MSELoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
    ```
  
  - **Training Loop**:
    ```python
    def train():
        model.train()
        optimizer.zero_grad()
        out = model(data)
        loss = criterion(out[data.train_mask].squeeze(), data.y[data.train_mask])
        loss.backward()
        optimizer.step()
        return loss.item()

    for epoch in range(1, 201):
        loss = train()
        if epoch % 10 == 0:
            print(f'Epoch {epoch}, Loss: {loss}')
    ```
  
- **Evaluate GNN Performance**:
  - **Define Evaluation Metrics**:
    - Mean Absolute Error (MAE)
    - Root Mean Squared Error (RMSE)
    - R-squared (optional)
  
  - **Evaluation Function**:
    ```python
    def test():
        model.eval()
        with torch.no_grad():
            out = model(data)
            pred = out[data.test_mask].squeeze()
            true = data.y[data.test_mask]
            mae = F.l1_loss(pred, true).item()
            rmse = torch.sqrt(F.mse_loss(pred, true)).item()
        return mae, rmse

    mae, rmse = test()
    print(f'Test MAE: {mae}, Test RMSE: {rmse}')
    ```
  
- **Save Trained Model**:
  - ```python
    torch.save(model.state_dict(), 'models/gcn_accessibility.pth')
    ```

- **Document GNN Implementation**:
  - Write detailed notes on model architecture, training parameters, and evaluation results.
  - Include diagrams if possible to illustrate the model structure.
